src/ast/irbuilderbpf.cpp

```diff
@@ -1669,143 +1644,156 @@ Value *IRBuilderBPF::CreateStrncmp(Value *str1,
-      else {
-        auto *ptr_l = CreateGEP(getInt8Ty(),
-                                CreatePointerCast(val1,
-                                                  getInt8Ty()->getPointerTo()),
-                                { getInt32(i + j) });
-        l = CreateLoad(getInt8Ty(), ptr_l);
-      }
-
-      Value *r;
-      if (literal2)
-        r = getInt8(literal2->c_str()[i]);
-      else {
-        auto *ptr_r = CreateGEP(getInt8Ty(),
-                                CreatePointerCast(val2,
-                                                  getInt8Ty()->getPointerTo()),
-                                { getInt32(i) });
-        r = CreateLoad(getInt8Ty(), ptr_r);
-      }
-
-      Value *cmp_null = CreateICmpEQ(r, null_byte, "strcontains.cmp_null");
-      CreateCondBr(cmp_null, done_true, cmp_char);
-
-      SetInsertPoint(cmp_char);
+  CreateBr(empty_check);
+  SetInsertPoint(empty_check);
+  Value *needle_first = CreateLoad(
+      getInt8Ty(), CreateGEP(getInt8Ty(), needle, { getInt64(0) }));
+  Value *needle_first_null = CreateICmpEQ(needle_first, null_byte);
+  CreateCondBr(needle_first_null, done_true, outer_cond);
+
+  SetInsertPoint(outer_cond);
+  Value *i_val = CreateLoad(getInt64Ty(), i);
+  Value *haystack_inbounds = CreateICmpULT(i_val, getInt64(haystack_sz));
+  CreateCondBr(haystack_inbounds, outer_cond_cmp, done_false);
+
+  SetInsertPoint(outer_cond_cmp);
+  Value *haystack_char = CreateLoad(
+      getInt8Ty(), CreateGEP(getInt8Ty(), haystack, { i_val }));
+  Value *haystack_valid = CreateICmpNE(haystack_char, null_byte);
+  CreateCondBr(haystack_valid, outer_body, done_false);
+
+  SetInsertPoint(outer_body);
+  CreateStore(getInt64(0), j);
+  CreateBr(inner_cond);
+
+  SetInsertPoint(inner_cond);
+  Value *j_val = CreateLoad(getInt64Ty(), j);
+  Value *needle_inbounds = CreateICmpULT(j_val, getInt64(needle_sz));
+  CreateCondBr(needle_inbounds, inner_body, outer_incr);
+
+  SetInsertPoint(inner_body);
+  Value *needle_char = CreateLoad(getInt8Ty(),
+                                  CreateGEP(getInt8Ty(), needle, { j_val }));
+  Value *needle_null = CreateICmpEQ(needle_char, null_byte);
+  CreateCondBr(needle_null, done_true, inner_notnull);
+
+  SetInsertPoint(inner_notnull);
+  Value *haystack_cmp_idx = CreateAdd(i_val, j_val);
+  Value *haystack_cmp_outbounds = CreateICmpUGE(haystack_cmp_idx,
+                                                getInt64(haystack_sz));
+  CreateCondBr(haystack_cmp_outbounds, outer_incr, inner_cmp);
+
+  // Inner conditional matching haystack char with needle char
+  SetInsertPoint(inner_cmp);
+  Value *haystack_cmp_char = CreateLoad(
+      getInt8Ty(), CreateGEP(getInt8Ty(), haystack, { haystack_cmp_idx }));
+  Value *haystack_cmp_needle = CreateICmpNE(haystack_cmp_char, needle_char);
+  Value *haystack_cmp_cond = CreateOr(haystack_cmp_outbounds,
+                                      haystack_cmp_needle);
+  CreateCondBr(haystack_cmp_cond, outer_incr, inner_incr);
```

Comment:
The `or` is no longer needed (already checked few lines above).
```suggestion
  CreateCondBr(haystack_cmp_needle, outer_incr, inner_incr);
```

Reply:
Oops, good catch